Optimizing K Framework Part 2/article.md

```diff
@@ -0,0 +1,552 @@
+
+To address this issue, we opted to [eliminate `UnifierT` from `NotSimplifier`](https://github.com/runtimeverification/haskell-backend/pull/3404) in the `runUnifierT` function, resulting in the following type signature:
+```hs
+runUnifierT :: MonadSimplify m => NotSimplifier m -> UnifierT m a -> m [a]
+```
+As a result of this modification, we needed to remove the `UnifierT` type variable from the definition of `NotSimplifier` in all other functions where `notSimplifier` was passed as an argument. Additionally, we [changed](https://github.com/runtimeverification/haskell-backend/pull/3387) `NotSimplifier` from a data type to a class to avoid the need for manual passing.
+
+After implementing the changes, the `trace-usage` tool showed that there were no longer any nested `UnifierT` types in our code. This enabled us to monomorphize and specialize the `simplifier` type variables, which can improve performance. The compiler can now inline code of this functions, leading to additional optimizations.
+
+## Monomorphization and specialization
+
+We used the `trace-usage` report to identify type variables that could be replaced with concrete types if only one type was instantiated during execution. For those whose type variables were instantiated to multiple different types, we added `SPECIALIZE` pragmas. We found that most functions could be monomorphized, and only a few required specialization. The process of monomorphization and specialization was divided into five pull requests: [part 1](https://github.com/runtimeverification/haskell-backend/pull/3336), [part 2](https://github.com/runtimeverification/haskell-backend/pull/3346), [part 3](https://github.com/runtimeverification/haskell-backend/pull/3375), [part 4](https://github.com/runtimeverification/haskell-backend/pull/3408), and [part 5](https://github.com/runtimeverification/haskell-backend/pull/3408).
+
+During this process, we also encountered cases of nested `MaybeT` and `ReaderT` caused by redundant `maybeT` and `runReaderT` calls, which added an unnecessary transformer layer in the argument. We resolved these issues by removing these redundant calls, which were straightforward solutions. We implemented these fixes in the pull requests titled [Remove nested ReaderT](https://github.com/runtimeverification/haskell-backend/pull/3418) and [Remove nested MaybeT](https://github.com/runtimeverification/haskell-backend/pull/3419).
+
+## Summary
+
+In this step we simplified the types in codebase by introducing a concrete `Simplifier` monad to replace the `SimplifierT` transformer. This was possible because we had unified `SMT` and `NoSMT` in previous article. We also addressed the issues with nested transformers, which allowed us to monomorphize most of the type variables that were constrained by `MonadSimplify`. This simplified the types and resulted in a speedup of approximately 15%.
```

Comment:
I would add that we also added `SPECIALIZE` pragmas where monomorphization wasn't possible.